staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/promise/promise_test.go

```diff
@@ -17,105 +17,103 @@ limitations under the License.
 	now := time.Now()
-	clock, counter := testclock.NewFake(now, 0, nil)
-	var lock sync.Mutex
-	wr := NewWriteOnce(&lock, counter)
-	var gots int32
-	var got atomic.Value
-	counter.Add(1)
-	go func() {
-		lock.Lock()
-		defer lock.Unlock()
-
-		got.Store(wr.Get())
-		atomic.AddInt32(&gots, 1)
-		counter.Add(-1)
-	}()
-	clock.Run(nil)
-	time.Sleep(time.Second)
-	if atomic.LoadInt32(&gots) != 0 {
-		t.Error("Get returned before Set")
-	}
-	func() {
-		lock.Lock()
-		defer lock.Unlock()
-		if wr.IsSet() {
-			t.Error("IsSet before Set")
-		}
-	}()
 	aval := &now
-	func() {
-		lock.Lock()
-		defer lock.Unlock()
-		if !wr.Set(aval) {
-			t.Error("Set() returned false")
-		}
-	}()
-	clock.Run(nil)
-	time.Sleep(time.Second)
-	if atomic.LoadInt32(&gots) != 1 {
-		t.Error("Get did not return after Set")
+	if !wr.Set(aval) {
+		t.Error("Set() returned false")
 	}
-	if got.Load() != aval {
-		t.Error("Get did not return what was Set")
+	expectGotValue(t, gots, aval)
+	goGetAndExpect(t, wr, gots, aval)
+	later := time.Now()
+	bval := &later
+	if wr.Set(bval) {
+		t.Error("second Set() returned true")
 	}
-	func() {
-		lock.Lock()
-		defer lock.Unlock()
-		if !wr.IsSet() {
-			t.Error("IsSet()==false after Set")
-		}
-	}()
-	counter.Add(1)
-	go func() {
-		lock.Lock()
-		defer lock.Unlock()
+	goGetAndExpect(t, wr, gots, aval)
+	cancel()
+	time.Sleep(3 * time.Second) // give it a chance to misbehave
+	goGetAndExpect(t, wr, gots, aval)
+}
 
-		got.Store(wr.Get())
-		atomic.AddInt32(&gots, 1)
-		counter.Add(-1)
-	}()
-	clock.Run(nil)
-	time.Sleep(time.Second)
-	if atomic.LoadInt32(&gots) != 2 {
-		t.Error("Second Get did not return immediately")
-	}
-	if got.Load() != aval {
-		t.Error("Second Get did not return what was Set")
+func TestWriteOnceCancel(t *testing.T) {
+	oldTime := time.Now()
+	cval := &oldTime
+	ctx, cancel := context.WithCancel(context.Background())
+	wr := NewWriteOnce(nil, ctx.Done(), cval)
+	gots := make(chan interface{})
+	goGetExpectNotYet(t, wr, gots, "cancel")
+	cancel()
+	expectGotValue(t, gots, cval)
+	goGetAndExpect(t, wr, gots, cval)
+	later := time.Now()
+	bval := &later
+	if wr.Set(bval) {
+		t.Error("Set() after cancel returned true")
 	}
-	func() {
-		lock.Lock()
-		defer lock.Unlock()
-		if !wr.IsSet() {
-			t.Error("IsSet()==false after second Get")
-		}
-	}()
+	goGetAndExpect(t, wr, gots, cval)
+}
+
+func TestWriteOnceInitial(t *testing.T) {
+	oldTime := time.Now()
+	cval := &oldTime
+	ctx, cancel := context.WithCancel(context.Background())
+	now := time.Now()
+	aval := &now
+	wr := NewWriteOnce(aval, ctx.Done(), cval)
+	gots := make(chan interface{})
+	goGetAndExpect(t, wr, gots, aval)
 	later := time.Now()
 	bval := &later
-	func() {
-		lock.Lock()
-		defer lock.Unlock()
-		if wr.Set(bval) {
-			t.Error("second Set() returned true")
-		}
+	if wr.Set(bval) {
+		t.Error("Set of initialized promise returned true")
+	}
+	goGetAndExpect(t, wr, gots, aval)
+	cancel()
+	time.Sleep(3 * time.Second) // give it a chance to misbehave
```

Comment:
To copy my comment as it disappeared:

12s is relatively long, especially given that our of those 12s, it's mostly sleeping. When I'm iterating with the code I would really like the tests to be as fast as possible.

Let me try to clarify what I meant:

if the imlementation is correct - it will never fail
if the implementation is incorrect (e.g. Get actually returns), then the longer timeout, the bigger probability of failure [it will probably not fail with 1us sleep in majority of cases, but will probably always fail with 1s]
What I'm saying is that in general, we make timeouts relatively long (waitTimeForever is 30s) to ensure that we won't get flakiness in heavy-loaded environments with huge cpu-starvation.
But that's not the case here - here you put 3s but it doesn't add any more value than what 1s would do - bigger timeouts won't help for potential flakiness.
In such case, I'm against artifically slowing down the tests, because that slows development and doesn't provide any real benefit.